src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -765,15 +777,37 @@ ExecuteSingleTask(QueryDesc *queryDesc, Task *task,
 		bool queryOK = false;
 		bool failOnError = false;
 		int64 currentAffectedTupleCount = 0;
-		PGconn *connection = GetConnectionForPlacement(taskPlacement,
-													   isModificationQuery);
+		int connectionFlags = SESSION_LIFESPAN;
+		MultiConnection *multiConnection = GetNodeConnection(connectionFlags,
+															 taskPlacement->nodeName,
+															 taskPlacement->nodePort);
+		PGconn *connection = multiConnection->pgConn;
+
+		if (taskRequiresTwoPhaseCommit)
+		{
+			RemoteTransactionBeginIfNecessary(multiConnection);
+			MarkRemoteTransactionCritical(multiConnection);
+		}
```

Comment:
>> and establish connections for all placements at once.

> Do you mean adding another loop before this?

Pretty much. Connection establishment and starting the transactions can be done entirely in parallel. Medium term we also should do what  #1032 suggests (i.e. sending the actual DML in parallel).